initrd/etc/luks-functions

```diff
@@ -267,24 +268,37 @@ prepare_thumb_drive()
 
 select_luks_container()
 {
+	#TODO: extend logic to prompt for block devices with model if multiple LUKS are found on block device instead of partitions
+	# Then feed luks with those partitions so that reencrypt and passphrase change can use passphrase to test all selected
 	TRACE_FUNC
 	if [ -s /boot/kexec_key_devices.txt ]; then
 		DEBUG "Reusing known good LUKS container device from /boot/kexec_key_devices.txt"
-		DEBUG "LUKS container device: $(cut -d ' ' -f1 /boot/kexec_key_devices.txt)"
-		LUKS=$(cut -d ' ' -f1 /boot/kexec_key_devices.txt)
-	else
+		LUKS=$(cut -d ' ' -f1 /boot/kexec_key_devices.txt) 
+		DEBUG "LUKS container device: $(echo $LUKS)"
+		# LUKS variable not exported yet, prompt for LUKS device
+	elif [ -z "$LUKS" ]; then
 		list_luks_devices > /tmp/luks_devices.txt
 		#if /tmp/luks_devices.txt exists and is not empty
 		if [ -s /tmp/luks_devices.txt ]; then
 			file_selector "/tmp/luks_devices.txt" "Select LUKS container device"
 			if [ "$FILE" == "" ]; then
 				return 1
 			else
-				LUKS=$FILE
-				detect_boot_device
-				mount -o remount,rw /boot
-				echo "$LUKS $(cryptsetup luksUUID $LUKS)" >/boot/kexec_key_devices.txt
-				mount -o remount,ro /boot
+				#TODO: What about BRTFS multi LUKS setup of QubesOS?
+				# if multiple LUKS containers are found on same block device
+				# select all of the luks containers on same block device instead of just one
+				# note that block devices for example under /dev/sda will be /dev/sda1, /dev/sda2, etc
+				# so we need to select all of the partitions on the same block device from /tmp/luks_devices.txt
+				# and then export them to LUKS variable
+				# then reencrypt and passphrase change functions will loop on all of the LUKS containers
+				# and test passphrase on all of them
+				if grep -q "$(echo $FILE | sed 's/[0-9]*$//')" /tmp/luks_devices.txt; then
+					DEBUG "Multiple LUKS containers found on same block device, selecting them all"
+					LUKS=$(grep $(echo $FILE | sed 's/[0-9]*$//') /tmp/luks_devices.txt)
```

Comment:
This seems very confusing and surprising that if you select one LUKS partition on a disk, it'll actually "automatically" alter the keys on all of the LUKS partitions on that disk.  This could be very unexpected if, say, you weren't actually using Qubes but you do have a second encrypted partition for whatever reason.

If that's what it has to do, we should ask the user to select a disk instead of a partition, perhaps listing all of the encrypted partitions on that disk since it will affect them all.

I'm not really sure that this is what it has to do, discussion above, it'll help to understand what exactly Qubes does with its LVM partitions and why we need to manipulate all of them (if we do).

Reply:
@JonathonHall-Purism I knew you wouldn't like this for BTRF 2 LUKS containers setup and file selector. Again, confusion present on default Q4.2 with single LUKS containing swap, dom0 rootfs lvm and other qubes lvms, which Heads doesn't care about. Heads just cares about LUKS containers. Not sure how to deal otherwise then current situation, exactly for your concern. Will then have to think more about this, but again, as per current code, expectation is for multiple LUKS containers to be unlockable with same Disk Recovery Key passphrase when setting DUK and reencrypting. A user having multiple LUKS containers would deploy manually, and should not have to reencrypt, and if he decides to set a DUK, he would select both LUKS to be decrypted automatically on boot, or only pass the partition linked to his LUKS FDE?

Comments?

Reply:
I'm more confused.  Is this trying to solve an issue with the default Qubes 4.2 partition layout or a custom layout?

OP currently says:
>cryptsetup2 toolstack version bump and script fixes to support multi-LUKS containers (BTRFS QubesOS 4.2) 

Which sounds like it's intending to solve a problem with the default setup for Qubes 4.2, but I don't know what a "multi-LUKS container" is (do you mean an LVM with mulitple logical volumes formatted with LUKS?), or what btrfs has to do with anything (we're just trying to unlock LUKS, not inspecting the root filesystem here)

This comment says:
>A user having multiple LUKS containers would deploy manually, and should not have to reencrypt, and if he decides to set a DUK, he would select both LUKS to be decrypted automatically on boot, or only pass the partition linked to his LUKS FDE?

So it is actually about a manual setup, not the default?  There are all sorts of manual setups possible.

Whenever I have had more than one LUKS partition (I do right now), I put an additional key in the root volume that's able to unlock the secondary volumes.  So I only unlock one password, and if I used a TPM DUK it'd still just need to unlock the root volume.  (I don't use that though :shrug: )

So again, the root question is, what partition setup are we talking about needing to support here?

Reply:
Wait, there's a comment above now answering my question I think, had to refresh the page (sorry, wish Github would either show updates or don't, really frustrating when it shows some of them...)

Reply:
OK, I get this now from the comment above.

Regarding the confusing UX, what would be wrong with prompting the user to select a disk instead of "prompt for partition, then assume they really want the whole disk and select all partitions on the disk"?

Reply:
@JonathonHall-Purism this was reworked. Now, all non-usb LUKS containers are tested against user provided LUKS DRK, and only unlockable LUKS with DRK passphrase are proposed to be used at once/user prompted to select only one LUKS container from that list for reencryption/passphrase change/TPM DUK. Good enough?

Reply:
Thanks.  This makes a lot more sense, UX is clearer and implementation is much easier to follow.  Looks good :+1: